features/lwipstack/LWIPInterfaceL3IP.cpp

```diff
@@ -0,0 +1,167 @@
+    if (netif.input(p, &netif) != ERR_OK) {
+        LWIP_DEBUGF(NETIF_DEBUG, ("Emac LWIP: IP input error\n"));
+
+        pbuf_free(p);
+    }
+}
+
+void LWIP::Interface::l3ip_state_change(bool up)
+{
+    if (up) {
+        tcpip_callback_with_block((tcpip_callback_fn)netif_set_link_up, &netif, 1);
+    } else {
+        tcpip_callback_with_block((tcpip_callback_fn)netif_set_link_down, &netif, 1);
+    }
+}
+
+#if LWIP_IGMP
+
+#include "lwip/igmp.h"
+/**
+ * IPv4 address filtering setup.
+ *
+ * \param[in] netif the lwip network interface structure
+ * \param[in] group IPv4 group to modify
+ * \param[in] action
+ * \return ERR_OK or error code
+ */
+err_t LWIP::Interface::l3ip_multicast_ipv4_filter(struct netif *netif, const ip4_addr_t *group, enum netif_mac_filter_action action)
+{
+    LWIP::Interface *mbed_if = static_cast<LWIP::Interface *>(netif->state);
+    SocketAddress addr(&group,NSAPI_IPv6);
+    switch (action) {
+        case NETIF_ADD_MAC_FILTER:
+        {
+            mbed_if->l3ip->add_ipv4_multicast_group(addr.get_ip_address());
+            return ERR_OK;
+        }
+        case NETIF_DEL_MAC_FILTER:
+        	mbed_if->l3ip->remove_ipv4_multicast_group(addr.get_ip_address());
+            return ERR_OK;
+        default:
+            return ERR_ARG;
+    }
+
+}
+#endif
+
+#if LWIP_IPV6_MLD
+
+#include "lwip/mld6.h"
+/**
+ * IPv6 address filtering setup.
+ *
+ * \param[in] netif the lwip network interface structure
+ * \param[in] group IPv6 group to modify
+ * \param[in] action
+ * \return ERR_OK or error code
+ */
+err_t LWIP::Interface::l3ip_multicast_ipv6_filter(struct netif *netif, const ip6_addr_t *group, enum netif_mac_filter_action action)
+{
+    LWIP::Interface *mbed_if = static_cast<LWIP::Interface *>(netif->state);
+    SocketAddress addr(&group,NSAPI_IPv6);
+    switch (action) {
+        case NETIF_ADD_MAC_FILTER:
+        {
+
+            mbed_if->l3ip->add_ipv6_multicast_group(addr.get_ip_address());
+            return ERR_OK;
+        }
+        case NETIF_DEL_MAC_FILTER:
+        	mbed_if->l3ip->remove_ipv6_multicast_group(addr.get_ip_address());
+            return ERR_OK;
+        default:
+            return ERR_ARG;
+    }
+
+}
+#endif
+
+err_t LWIP::Interface::l3ip_if_init(struct netif *netif)
+{
+    int err = ERR_OK;
+    LWIP::Interface *mbed_if = static_cast<LWIP::Interface *>(netif->state);
+
+    mbed_if->l3ip->set_memory_manager(*mbed_if->memory_manager);
+    mbed_if->l3ip->set_link_input_cb(mbed::callback(mbed_if, &LWIP::Interface::l3ip_input));
+    mbed_if->l3ip->set_link_state_cb(mbed::callback(mbed_if, &LWIP::Interface::l3ip_state_change));
+
+    /* Interface capabilities */
+    netif->flags = NETIF_FLAG_BROADCAST | NETIF_FLAG_ETHARP | NETIF_FLAG_ETHERNET;
+
+    if (!mbed_if->l3ip->power_up()) {
+        err = ERR_IF;
+    }
+
+    netif->mtu = mbed_if->l3ip->get_mtu_size();
+    mbed_if->l3ip->get_ifname(netif->name, 2);
```

Comment:
If this is used for socket bind to interface name, length needs be longer than two characters in future. So needs change to lwip internal data structures. But that is not needed now in this first commit, so leave this as it is, for now.

Reply:
fixed